hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/validation/ValidatorRegistry.java

```diff
@@ -0,0 +1,201 @@
+import org.reflections.Reflections;
+import org.reflections.scanners.MethodAnnotationsScanner;
+import org.reflections.util.ClasspathHelper;
+import org.reflections.util.ConfigurationBuilder;
+
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Registry that loads and stores the request validators to be applied by
+ * a service.
+ */
+public class ValidatorRegistry {
+
+  private final EnumMap<
+      ValidationCondition, EnumMap<Type, Pair<List<Method>, List<Method>>>>
+      validators = new EnumMap<>(ValidationCondition.class);
+
+  /**
+   * Creates a {@link ValidatorRegistry} instance that discovers validation
+   * methods in the provided package and the packages in the same resource.
+   * A validation method is recognized by the {@link RequestFeatureValidator}
+   * annotation that contains important information about how and when to use
+   * the validator.
+   * @param validatorPackage the main package inside which validatiors should
+   *                         be discovered.
+   */
+  ValidatorRegistry(String validatorPackage) {
+    this(ClasspathHelper.forPackage(validatorPackage));
+  }
+
+  /**
+   * Creates a {@link ValidatorRegistry} instance that discovers validation
+   * methods under the provided URL.
+   * A validation method is recognized by the {@link RequestFeatureValidator}
+   * annotation that contains important information about how and when to use
+   * the validator.
+   * @param searchUrls the path in which the annotated methods are searched.
+   */
+  ValidatorRegistry(Collection<URL> searchUrls) {
+    Reflections reflections = new Reflections(new ConfigurationBuilder()
+        .setUrls(searchUrls)
+        .setScanners(new MethodAnnotationsScanner())
+        .useParallelExecutor()
+    );
+
+    Set<Method> describedValidators =
+        reflections.getMethodsAnnotatedWith(RequestFeatureValidator.class);
+    initMaps(describedValidators);
+  }
+
+  /**
+   * Get the validators that has to be run in the given list of
+   * {@link ValidationCondition}s, for the given requestType and
+   * {@link RequestProcessingPhase}.
+   *
+   * @param conditions conditions that are present for the request
+   * @param requestType the type of the protocol message
+   * @param phase the request processing phase
+   * @return the list of validation methods that has to run.
+   */
+  List<Method> validationsFor(
+      List<ValidationCondition> conditions,
+      Type requestType,
+      RequestProcessingPhase phase) {
+
+    if (conditions.isEmpty() || validators.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    Set<Method> returnValue =
+        new HashSet<>(validationsFor(conditions.get(0), requestType, phase));
```

Comment:
You can define the same validator to be used for multiple conditions, as the conditions field in the annotation is defined to be an array. In this case the validator will be added into the list inside both condition for the request type and processing phase.

Also when you are requesting for the validation via this method, you can specify multiple conditions, as the cluster can be in a pre-finalized state when an older client issues a request, and in this case if the same validation is set for both conditions, it would be returned twice, and then the validator would run twice which might be either unnecessary, unexpected or misbehaving the second time.

As you might ask for a use case, I do not have one in mind, but having this in does not seem to be a big price for having this flexibility, though if you think, I am open to drop this, and we can change the conditions to be condition, and let just one condition to be specified in the annotation.

Reply:
Makes sense, thanks for the explanation.